pgx/src/stringinfo.rs

```diff
@@ -65,51 +59,64 @@ impl std::io::Write for StringInfo {
     }
 }
 
-impl ToString for StringInfo {
+impl<AllocatedBy: WhoAllocated> ToString for StringInfo<AllocatedBy> {
+    /// Convert this [`StringInfo`] into a Rust string.  This uses [`String::from_utf8_lossy`] as
+    /// it's fine for a Postgres [`StringInfo`] to contain null bytes and also not even be proper
+    /// UTF8.
     fn to_string(&self) -> String {
-        unsafe { std::str::from_utf8_unchecked(self.as_bytes()).to_owned() }
+        String::from_utf8_lossy(self.as_bytes()).to_string()
     }
 }
 
-impl StringInfo {
-    /// Construct a new `StringInfo` of its default size, allocated by Postgres in `CurrentMemoryContext`
+impl StringInfo<AllocatedByRust> {
+    /// Construct a new `StringInfo` of its default size, allocated in `CurrentMemoryContext`
     ///
-    /// Unless `.into_pg()` or `.into_char_ptr()` are called, memory management of
-    /// this `StringInfo` follow Rust's drop semantics.
+    /// Note that Postgres can only represent up to 1 gigabyte of data in a `StringInfo`
     pub fn new() -> Self {
-        StringInfo { sid: unsafe { pg_sys::makeStringInfo() }, needs_pfree: true }
+        StringInfo {
+            inner: unsafe {
+                // SAFETY: makeStringInfo() always returns a valid StringInfoData pointer.  It'll
+                // ereport if it can't
+                PgBox::<_, AllocatedByRust>::from_rust(pg_sys::makeStringInfo())
+            },
+        }
     }
 
     /// Construct a new `StringInfo`, allocated by Postgres in `CurrentMemoryContext`, ensuring it
     /// has a capacity of the specified `len`.
     ///
     /// Note that Postgres can only represent up to 1 gigabyte of data in a `StringInfo`
-    ///
-    /// Unless `.into_pg()` or `.into_char_ptr()` are called, memory management of
-    /// this `StringInfo` follow Rust's drop semantics.
-    pub fn with_capacity(len: usize) -> Self {
+    pub fn with_capacity(len: i32) -> Self {
```

Comment:
Hm, why the change to the `i32` type?

Reply:
I argued with myself over this.  Postgres' `StringInfo` length is an `i32`, so I thought rather than us deal with an `usize` that overflows an i32 and have to return a Result, just let the user sort it out ahead of time.

Thoughts?

Reply:
We discussed this on Discord and my sentiment is broadly "`usize` is fine and indicates what the input is "for" to Rust programmers (morally it's actually `pg_size_t`), and Postgres is running so many checks that it will throw a fit if we throw a bad input into Postgres, don't worry about it". `with_capacity` is really more of an "optimization hint" to Postgres, as Postgres will, before pushing anything on StringInfo, always calculate `needed` before pushing data into it.